kroxylicious-filters/kroxylicious-encryption/src/test/java/io/kroxylicious/filter/encryption/inband/RecordEncryptorTest.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption.inband;
+
+import java.nio.ByteBuffer;
+import java.util.Set;
+
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.internals.RecordHeader;
+import org.apache.kafka.common.record.Record;
+import org.apache.kafka.common.record.RecordBatch;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+import io.kroxylicious.filter.encryption.EncryptionScheme;
+import io.kroxylicious.filter.encryption.EncryptionVersion;
+import io.kroxylicious.filter.encryption.RecordField;
+import io.kroxylicious.test.assertj.KafkaAssertions;
+import io.kroxylicious.test.record.RecordTestUtils;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+
+class RecordEncryptorTest {
```

Comment:
I'd be keen to see the unit tests checking that the Parcel buffer handed to the KeyContext has the expected contents, and also check the Wrapper bytes are constructed correctly. Rather than the valueNotEqualTo checks.

Reply:
I agree, but it's tricky to really prove that the structure is right (not least because although we have `Parcel`, we don't have a corresponding `Wrapper` at the moment). I wrote the corresponding `RecordDecryptor`, which doesn't prove exactly the same thing, does prove that the mutual contract between these classes holds.

Reply:
cool, happy enough with the roundtrip testing for now. 

I wonder if it would be good in future to have test code to independently verify the structure though, for example we make changes that keep things internally consistent between RecordEncryptor and RecordDecryptor but break the ability to decrypt old parcels. Or it could start generating V1 parcels that cannot be decoded by an older version of the proxy that should in theory be able to read V1.

Maybe that means we need some solid tests on Parcel/Wrapper demonstrating we can decrypt old parcels from binary, though I imagine a directory with binary examples also not being very maintainable.

Not sure how we check that we are schema compliant for older clients, I guess we would need test code that can decrypt parcels and is aware of the schema. A lot of code to maintain.

Reply:
having a binary format like avro/protobuf in the mix could help with the schema aspect, is there something we could take from the kafka client around schemas?

Reply:
> I wonder if it would be good in future to have test code to independently verify the structure though

Nothing against that. Once we've migrated the existing code to use `RecordEncryptor` and `RecordDecryptor` and use the `DekManager` I intend to circle back to the Parcel and Wrapper code and rework it so that we only need a single buffer, which I believe should be possible.

> break the ability to decrypt old parcels

The way to do that is to some "old" wrappers in files and have tests that we can decrypt them.

> though I imagine a directory with binary examples also not being very maintainable

Maybe not, but I think it's the only way to be sure.

> having a binary format like avro/protobuf in the mix could help with the schema aspect, is there something we could take from the kafka client around schemas?

Kafka has some things, but not as public API. I'd rather not pick up a dependency on anything else just for this. It's not a huge amount of code.